ssi-sd-jwt/src/decode.rs

```diff
@@ -0,0 +1,235 @@
+use serde::de::DeserializeOwned;
+use serde::Deserialize;
+use ssi_jwk::JWK;
+use ssi_jwt::NumericDate;
+use std::collections::BTreeMap;
+
+use crate::serialized::deserialize_string_format;
+use crate::verify::{DecodedDisclosure, DisclosureKind};
+use crate::*;
+
+/// Expiration validity claims that are sampled before expanding selective disclosures
+#[derive(Debug, Deserialize, PartialEq)]
+pub struct ValidityClaims {
+    /// Not Before claim
+    pub nbf: Option<NumericDate>,
+
+    /// Issued After claim
+    pub iat: Option<NumericDate>,
+
+    /// Expiration claim
+    pub exp: Option<NumericDate>,
+}
+
+/// High level API to decode a fuilly encoded SD-JWT.  That is a JWT and selective
+/// disclosures seperated by tildes
+pub fn decode_verify<Claims: DeserializeOwned>(
+    serialized: &str,
+    key: &JWK,
+) -> Result<(ValidityClaims, Claims), DecodeError> {
+    let deserialized = deserialize_string_format(serialized)
+        .ok_or(DecodeError::UnableToDeserializeStringFormat)?;
+
+    decode_verify_disclosure_array(deserialized.jwt, key, &deserialized.disclosures)
+}
+
+/// Lower level API to decode an SD-JWT that has already been split into it's
+/// JWT and disclosure components
+pub fn decode_verify_disclosure_array<Claims: DeserializeOwned>(
+    jwt: &str,
+    key: &JWK,
+    disclosures: &[&str],
+) -> Result<(ValidityClaims, Claims), DecodeError> {
+    let mut payload_claims: serde_json::Value = ssi_jwt::decode_verify(jwt, key)?;
+
+    let validity_claims: ValidityClaims = serde_json::from_value(payload_claims.clone())?;
+
+    let sd_alg = sd_alg(&payload_claims)?;
+    let _ = payload_claims
+        .as_object_mut()
+        .unwrap()
+        .remove(SD_ALG_CLAIM_NAME);
+
+    let mut disclosures = translate_to_in_progress_disclosures(disclosures, sd_alg)?;
+
+    visit_claims(&mut payload_claims, &mut disclosures)?;
+
+    for (_, disclosure) in disclosures {
+        if !disclosure.found {
+            return Err(DecodeError::UnusedDisclosure);
+        }
+    }
+
+    Ok((validity_claims, serde_json::from_value(payload_claims)?))
+}
+
+fn sd_alg(claims: &serde_json::Value) -> Result<SdAlg, DecodeError> {
+    let alg_name = claims[SD_ALG_CLAIM_NAME]
+        .as_str()
+        .ok_or(DecodeError::MissingSdAlg)?;
+
+    SdAlg::try_from(alg_name)
+}
+
+fn translate_to_in_progress_disclosures(
+    disclosures: &[&str],
+    sd_alg: SdAlg,
+) -> Result<BTreeMap<String, InProgressDisclosure>, DecodeError> {
+    let disclosure_vec: Result<Vec<_>, DecodeError> = disclosures
+        .iter()
+        .map(|disclosure| InProgressDisclosure::new(disclosure, sd_alg))
+        .collect();
+
+    let disclosure_vec = disclosure_vec?;
+
+    let mut disclosure_map = BTreeMap::new();
+    for disclosure in disclosure_vec {
+        let prev = disclosure_map.insert(disclosure.hash.clone(), disclosure);
+
+        if prev.is_some() {
+            return Err(DecodeError::MultipleDisclosuresWithSameHash);
```

Comment:
I have mixed feelings about this returning an error rather than just warning or ignoring. There are two reasons why you might have multiple disclosures with the same hash:

1) User/library error, included the same disclosure twice.
2) An attacker is performing a hash collision attack (let's ignore the feasibility of the attack), and they include both the genuine and fake elements in the disclosure list.

Since we would only be protecting against a clumsy attacker who forgot to remove the genuine element from the disclosure list, I think it would be safe to assume that 1) would account for the vast majority of cases. For that reason I would be in favour of one of the following:

a) Ignoring the duplicates (would simplify the code too as you would only need one iteration through the elements, collecting directly into a map).
b) Warning instead of erroring if 2) is still a concern.

Reply:
This appears to be another spec compliance issue

> If any digests were found more than once in the previous step, the SD-JWT MUST be rejected.

https://drafts.oauth.net/oauth-selective-disclosure-jwt/draft-ietf-oauth-selective-disclosure-jwt.html#section-6.1

which, when combined with the restriction above saying that all disclosures must be referenced, my reading is that each disclosure must have a unique hash.

Reading a little between the lines, I believe the thinking of the spec is that, given complex JWT constructions like array style disclosures, nested SD-JWTs, and recursive SD-JWTs, it is possible to otherwise construct an SD-JWT where the same disclosure does come up in multiple places in the same JWT.  Because of how easy it is for an implementation to dump the hashes into an associative structure, I think they wanted to ban duplicate hashes in general to take a lot of goofy constructions off the table.  Because of the salts in disclosures, you can duplicate the same data in multiple claim tree nodes; they simply need to be in their own disclosures instead of doubling up disclosures with the same value.

Reply:
I understand the spec differently here. My understanding is if the SD-JWT contains a digest more than once (i.e. once in an array element and once in a "_sd" array), then the SD-JWT is considered invalid. But I think the spec is unclear here, so happy to go with the strict interpretation for now and we can get clarification.